servers/rendering/renderer_rd/renderer_canvas_render_rd.cpp

```diff
@@ -2747,36 +2822,297 @@ void RendererCanvasRenderRD::_record_item_commands(const Item *p_item, RenderTar
 	{
-		RD::Uniform u_diffuse(RD::UNIFORM_TYPE_TEXTURE, 0, p_batch->tex_info.diffuse);
-		RD::Uniform u_normal(RD::UNIFORM_TYPE_TEXTURE, 1, p_batch->tex_info.normal);
-		RD::Uniform u_specular(RD::UNIFORM_TYPE_TEXTURE, 2, p_batch->tex_info.specular);
-		RD::Uniform u_sampler(RD::UNIFORM_TYPE_SAMPLER, 3, p_batch->tex_info.sampler);
-		RD::Uniform u_instance_data(RD::UNIFORM_TYPE_STORAGE_BUFFER, 4, state.canvas_instance_data_buffers[state.current_data_buffer_index].instance_buffers[p_batch->instance_buffer_index]);
+		RIDSetKey key(
+				&p_batch->batch_textures[0],
+				p_batch->batch_textures_used,
+				&p_batch->batch_samplers[0],
+				p_batch->batch_samplers_used,
+				state.canvas_instance_data_buffers[state.current_data_buffer_index].instance_buffers[p_batch->instance_buffer_index],
+				state.canvas_instance_data_buffers[state.current_data_buffer_index].texture_data_buffers[state.current_texture_data_buffer_index]);
+
+		const RID *uniform_set = rid_set_to_uniform_set.getptr(key);
+		if (uniform_set == nullptr) {
+			{
+				RID *ptr = state.texture_data_uniform_texture_rids.ptrw();
+				for (uint32_t i = 0; i < batch_available_texture_slots; i++) {
+					if (p_batch->batch_textures[i].is_valid()) {
+						ptr[i] = p_batch->batch_textures[i];
+					} else {
+						ptr[i] = default_texture_info.diffuse;
+					}
+				}
+			}
+
+			{
+				RID *ptr = state.texture_data_uniform_sampler_rids.ptrw() + DEFAULT_MATERIAL_SAMPLER_COUNT;
+				for (uint32_t i = 0; i < BATCH_MAX_DYNAMIC_SAMPLERS; i++) {
+					if (p_batch->batch_samplers[i].is_valid()) {
+						ptr[i] = p_batch->batch_samplers[i];
+					} else {
+						ptr[i] = default_texture_info.sampler;
+					}
+				}
+			}
+
+			state.batch_texture_uniforms.write[0] = RD::Uniform(RD::UNIFORM_TYPE_TEXTURE, 0, state.texture_data_uniform_texture_rids);
+			state.batch_texture_uniforms.write[1] = RD::Uniform(RD::UNIFORM_TYPE_SAMPLER, 1, state.texture_data_uniform_sampler_rids);
+			state.batch_texture_uniforms.write[2] = RD::Uniform(RD::UNIFORM_TYPE_STORAGE_BUFFER, 2, state.canvas_instance_data_buffers[state.current_data_buffer_index].instance_buffers[p_batch->instance_buffer_index]);
+			state.batch_texture_uniforms.write[3] = RD::Uniform(RD::UNIFORM_TYPE_STORAGE_BUFFER, 3, state.canvas_instance_data_buffers[state.current_data_buffer_index].texture_data_buffers[state.current_texture_data_buffer_index]);
+
+			RID rid = RD::get_singleton()->uniform_set_create(state.batch_texture_uniforms, shader.default_version_rd_shader, BATCH_UNIFORM_SET);
+			ERR_FAIL_COND_MSG(rid.is_null(), "Failed to create uniform set for batch.");
 
-		RID uniform_set = uniform_set_cache->get_cache(shader.default_version_rd_shader, BATCH_UNIFORM_SET, u_diffuse, u_normal, u_specular, u_sampler, u_instance_data);
+			const RIDCache::Pair *iter = rid_set_to_uniform_set.insert(std::move(key.to_owned()), rid);
```

Comment:
Where is `std::move` getting included from? I thought it came from `<utility>` but that seems to only be included in thirdparty code. 

Ideally, we would do without any std functions unless absolutely necessary. Is there any way to replace this with something that doesn't depend on the STD?

Reply:
`static_cast<T&&>`